### PR TITLE
Charm++ for Shasta and namd with plumed.

### PR DESCRIPTION
**Pull request by halfflat**
* Add patches to allow charm++ to run over OFI backend on Shasta.
* Extend plumed recipe to allow force patching an engine (the safety test included in plumed's lib/plumed/patches/named-2.14.config checks for the existence of a file */Make.config, but none such are present in the namd-2.14 source tree).
* Patch namd in turn to build against this version of charm++.
* Patch namd to work with recent pythons (need `--embed` now to get library paths).
* Add plumed variant for namd that applies the plumed patches.

### CODE REVIEW

File: setonix/environments/env_apps/spack.yaml [hunk 1]
```diff
@@ -19,23 +19,14 @@ spack:
...
       +user-uef +user-yaff ^kokkos@3.4.01+hwloc+memkind+numactl+openmp+tuning ^plumed@2.7.2
-
-    
-# TODO: charmpp+namd yet to be made to work with Shasta/Slingshot
-    - namd@2.14 interface=python fftw=3 ^charmpp@7.0.0
-      backend=mpi
-
-    
+      ^python@3.9.7+optimizations
+    - namd@2.14 interface=python fftw=3 +plumed ^python@3.9.7+optimizations ^charmpp@6.10.2 backend=ofi
```
Comment (marcodelapierre) [inline setonix/environments/env_apps/spack.yaml:23]: @halfflat you can get rid of the occurrences of `^python@3.9.7+optimizations`, it is not enforced in `packages.yaml` once and for all packages :)

**Review by marcodelapierre — APPROVED** [review]
Thanks Sam, very good to me!

Just a minor comment on the spack.yaml, but please go ahead!
For all of this work which is not Pawsey-hardware-specific (most of it I think?), consider aiming for a PR in the main Spack public repo :)

File: setonix/repo_setonix/packages/plumed/package.py [hunk 2]
```diff
@@ -91,7 +91,7 @@ class Plumed(AutotoolsPackage):
 
     parallel = False
 
-    def apply_patch(self, other):
+    def apply_patch(self, other, force=False):
```
Comment (pelahi) [inline setonix/repo_setonix/packages/plumed/package.py:94]: Any reason for the force=False?

File: setonix/environments/env_apps/spack.yaml (hunk 1 above)
Comment (pelahi) [reply setonix/environments/env_apps/spack.yaml:23]: I agree with @marcodelapierre.

**Review by pelahi — APPROVED** [review]
change the `^python@3.9.7 +optimizations` and I think this looks fine.

File: setonix/repo_setonix/packages/plumed/package.py (hunk 2 above)
Comment (halfflat) [reply setonix/repo_setonix/packages/plumed/package.py:94]: Yep, this is to cover the default case (where we probably do want to keep the plumed patch check) as it applies in other recipes, such as the one for gromacs.

**Pull request merged**